### PR TITLE
Add time elapsed/remaining to effects info panel

### LINKED ISSUE #48770: View with estimated duration of pills taken
# Is your feature request related to a problem? Please describe.

Based on food handling skill we can estimate food's time to expire.
We could add view where player could see history of taken drugs and medicine either in form of "X time remaining" or "taken X time ago"

# Describe the solution you'd like

I don't have good idea where to put it UI-wise.
Current `@` is already very clustered, making another window just for that seems forced, unless some other things would be moved there.

Potentially we could make extended effect view separated on ailments and group of possible solutions:
on left side Weakness, on the right an option to take pills or use Radiation Scrubber
on the left "headache" on the right an option to take antiparasitics
on the left Infection, on the right info "you took strong antibiotic 3 hours ago"
Being able to see solutions on the right would depend on player's healthcare skill

Optionally, under `@` in effects tab, there could be information how long ago effect started.

# Describe alternatives you've considered

Some item/Self-Aware mutation/device being able to log those information, the way Self Aware allows you to see your `health`.

### PR DESCRIPTION
**Pull request by WelpThatWorked**
#### Summary
Interface "Add time elapsed to effects info panel"

#### Purpose of change
Fix #48770 

#### Describe the solution
Add time since start of an effect to effect info panel. Add time remaining, effect ID and effect intensity to info panel if debug mode on. 
Moved estimate_effect_dur() from avatar to Character class, and changed implementation. Previously it used rng() rather than rng_float(), making it only capable of returning results with zero or maximum error. Add minimum error value allowing a minimum spread even if the skill threshold is exceeded. 

#### Describe alternatives you've considered
None

#### Testing
Gave self several different effects of varying length and intensity. 

#### Additional context
Example with debug mode ON
image

### CODE REVIEW

**Comment by Venera3** [comment]
I'm not very keen on the player magically getting this info down to the second, regardless of healthcare skill. Even the known, pharmacy-grade substances have a high variance in their work duration IRL, and the effect system is used for a lot of things you would never be able to guesstimate beyond what the intensity levels already tell you (ex. bleeding, pain, completely unknown venoms). The fact that the current effect durations are completely deterministic is mostly just a limitation of the system. 

I could see it as a more readable component of Debug mode, though, especially if you hooked it up to the mob examine window and did away with the current message log spam. Alternatively, you could limit it to completely known drugs and limit the best precision to the level of "about an hour left".

**Comment by actual-nh** [comment]
> I'm not very keen on the player magically getting this info down to the second, regardless of healthcare skill. Even the known, pharmacy-grade substances have a high variance in their work duration IRL,

Part of that is variation between people (and between whether someone's on various other medications, etc - cytochrome p450 interactions are something I have to check on for myself a lot...), but I have to agree on an hour being the most precision possible for most oral drugs (with a potentially noticeable effect) for one person (from personal experience).

> and the effect system is used for a lot of things you would never be able to guesstimate beyond what the intensity levels already tell you (ex. bleeding, pain, completely unknown venoms). The fact that the current effect durations are completely deterministic is mostly just a limitation of the system.

Agreed.

> I could see it as a more readable component of Debug mode, though, especially if you hooked it up to the mob examine window and did away with the current message log spam. Alternatively, you could limit it to completely known drugs and limit the best precision to the level of "about an hour left".

Or both? (What are your thoughts regarding the self-aware trait/mutation/whatever, BTW?)

**Comment by Saicchi** [comment]
iirc Self-Aware trait is set to be removed in the future, so I don't think adding more to it is a good idea.

**Comment by WelpThatWorked** [comment]
The "Limit to known drugs" thing was something I thought about, but there's currently no way to determine if a particular effect has a known length or not (would have to be specified in the effects list of the consumable, as some effects might come from multiple sources?). 
I agree it's probably best to drop the Time remaining function altogether, I don't feel great about how it works and it doesn't feel very realistic, mostly for the reasons you describe. 
I'll try to do the Debug mode thing though.

**Comment by Saicchi** [comment]
> I'll try to do the Debug mode thing though.
Just in case you don't know.
`debug_mode` is a global variable from `debug.h` which enables / disables debug mode.

**Comment by WelpThatWorked** [comment]
> Just in case you don't know.
> `debug_mode` is a global variable from `debug.h` which enables / disables debug mode.

Got that, thanks. 
Should I remove the existing debug messages or leave them?

**Comment by actual-nh** [comment]
> Should I remove the existing debug messages or leave them?

After testing, I'd probably comment them out.

File: src/effect.cpp [hunk 1]
```diff
@@ -972,12 +997,14 @@ int effect::set_intensity( int val, bool alert )
...
 
-    int old_intensity = intensity;
+    //int old_intensity = intensity;
     intensity = val;
+    /*
     if( old_intensity != intensity ) {
         add_msg_debug( debugmode::DF_EFFECT, "%s intensity %d->%d", get_id().c_str(), old_intensity,
                        intensity );
     }
+    */
```
Comment (kevingranade) [inline src/effect.cpp:1007]:
```suggestion
    intensity = val;
```

File: src/effect.cpp [hunk 2]
```diff
@@ -40,6 +40,7 @@ static const itype_id itype_money_bundle( "money_bundle" );
 
 static const trait_id trait_LACTOSE( "LACTOSE" );
 static const trait_id trait_VEGETARIAN( "VEGETARIAN" );
+static const trait_id trait_SELFAWARE( "SELFAWARE" );
```
Comment (kevingranade) [inline src/effect.cpp:43]:
```suggestion
```

**Pull request merged**